### ISSUE TITLE
Engintron gives 400 after disabling and then re-enabling it.

### ISSUE THREAD

**Issue by lpasquali** [engintron/engintron#973]
Hello
I'm quite confused.
Cpanel and engintron versions affected for us:

cpanel 74.0.10 and cpanel 74.0.11  with cloudlinux 7.5
Engintron 1.91 and 1.9.2

How to reproduce:
- machine with any cPanel version in 74.0.x (it's not a good time for us to switch to 76.x)
- last or previous version of engintron

The cPanel server testing url apache only gives a response like that:
```
vanth:~ luca$ curl -I http://monitoir.example.com/monitor.php
HTTP/1.1 200 OK
Date: Wed, 05 Dec 2018 16:48:14 GMT
Server: Apache
X-Powered-By: PHP/7.1.23
Upgrade: h2,h2c
Connection: Upgrade
Vary: User-Agent
Content-Type: text/html; charset=UTF-8
vanth:~ luca$
```

when nginx gets enabled first time:
```
root@demo11 luca]# curl -I http://monitoir.example.com/monitor.php
HTTP/1.1 200 OK
Server: nginx
Date: Wed, 05 Dec 2018 16:34:36 GMT
Content-Type: text/html; charset=UTF-8
Connection: keep-alive
Vary: Accept-Encoding
X-Powered-By: PHP/7.1.23
Vary: Accept-Encoding,User-Agent
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
X-Nginx-Cache-Status: EXPIRED
X-Server-Powered-By: Engintron
[root@demo11 luca]#
```
after nginx is disabled the output is similar to the first one pasted, as apache is all alone again.

after nginx is re-enabled:
```
[root@demo11 luca]# curl -I http://monitoir.example.com/monitor.php
HTTP/1.1 400 Bad Request
Server: nginx
Date: Wed, 05 Dec 2018 16:50:25 GMT
Content-Type: text/html
Content-Length: 242
Connection: keep-alive
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
[the last two lines get repeated around 500 times each]
```

for some reason https work fine:
```
vanth:~ luca$ curl -I I https://monitoir.example.com/monitor.php
HTTP/2 200 
server: nginx
date: Wed, 05 Dec 2018 17:07:52 GMT
content-type: text/html; charset=UTF-8
vary: Accept-Encoding
x-powered-by: PHP/7.1.23
vary: Accept-Encoding,User-Agent
x-xss-protection: 1; mode=block
x-content-type-options: nosniff
x-nginx-cache-status: BYPASS
x-server-powered-by: Engintron
vanth:~ luca$ 
```

**Comment by fevangelou** [comment]
This is obviously system-specific. Check for rogue .htaccess redirects going into loops.

Alternatively, just google it, e.g. this could be the case: https://stackoverflow.com/questions/12315832/how-to-fix-nginx-throws-400-bad-request-headers-on-any-header-testing-tools

**Comment by lpasquali** [comment]
Dear Fotis,

Sorry but I think you misread this report, I admit I was a lot verbose in the last two days, my bad too. First of all, .htaccess is not responsible for anything, the url used with curl is a default account for monitoring with a stupid php script which gives "Up" as an output, which is parsed by Pingdom for example. I can share any cpanel configuration, no problem, but besides cPanel, something bad happens in the vhost generation after I take engintron down and the up again, I described you a first installation, I'm going to test just for couriosity if this happens also on cPanel 76.x.

by the way if I request directly apache, from the machine I get:
```
[luca@demo11 ~]$  curl -I http://monitor-demo11.netsons.net:8080/monitor.php
HTTP/1.1 200 OK
Date: Thu, 06 Dec 2018 08:55:12 GMT
Server: Apache
X-Powered-By: PHP/7.1.23
Upgrade: h2,h2c
Connection: Upgrade
Vary: User-Agent
Content-Type: text/html; charset=UTF-8

[luca@demo11 ~]$ 

[luca@demo11 ~]$  curl -I http://monitor-demo11.netsons.net/monitor.php
HTTP/1.1 400 Bad Request
Server: nginx
Date: Thu, 06 Dec 2018 08:56:30 GMT
Content-Type: text/html
Content-Length: 242
Connection: keep-alive
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
[.. the two lines keep repeating 500 times each ..] 
```

also this test takes lsapi php engine out of the discussion, file pippo.txt has just work "pippo" inside:

[root@demo11 luca]# curl -I http://monitor-demo11.netsons.net/pippo.txt
HTTP/1.1 400 Bad Request
Server: nginx
Date: Thu, 06 Dec 2018 09:02:41 GMT
Content-Type: text/html
Content-Length: 242
Connection: keep-alive
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
[ .. same as above ..]

[root@demo11 luca]# curl -I http://monitor-demo11.netsons.net:8080/pippo.txt
HTTP/1.1 200 OK
Date: Thu, 06 Dec 2018 09:04:10 GMT
Server: Apache
Upgrade: h2,h2c
Connection: Upgrade
Last-Modified: Thu, 06 Dec 2018 09:01:19 GMT
Accept-Ranges: bytes
Content-Length: 6
Vary: User-Agent
Content-Type: text/plain

[root@demo11 luca]# 

and just for clarification:

```
[root@demo11 luca]# ls -la /home/demo11/public_html/
total 8
drwxr-x---  4 demo11 nobody   59 Nov 16 23:37 .
drwx--x--x 17 demo11 demo11 4096 Nov 19 12:18 ..
drwxr-xr-x  2 demo11 demo11    6 Nov 16 23:37 cgi-bin
-rw-r--r--  1 demo11 demo11  173 Nov 16 23:37 monitor.php
drwxr-xr-x  4 demo11 demo11   50 Nov 16 23:38 .well-known
[root@demo11 luca]# 
```

How is possible that this account has something local  that breaks up http only?
I also see that the loop which comes after the error 400 in the curl headers, looks like to be related to this piece of configuration for the  /etc/nginx/nginx.conf you maintain:

    # Security Headers
    add_header        X-XSS-Protection "1; mode=block" always;
    add_header        X-Content-Type-Options "nosniff" always; 

Our cPanel tweaking is mostly focused on packages resources per product line sold, concerning engintron we customized rules in this logic all configurations go in custom_rules and we just modify resolvers to point to us for any dns resolution (which should be not much as we specify the cpanel ip address to proxy for each domain/vhost):
- anyone has the proxy variable set to its correct ip address so if cloudflare is enabled, it works, apache logging has been fixed for that.
- cache for static and dynamic content is disabled by default, we dynamically include configuration bits with a vhost list (a regular expression which also works if user sets up wildcards)) per cpanel account, which configures correct ip address.

Please read this carefully, you can see that there is something wrong in the configuration or in what is generated from cPanel in the vhosts config file.

Sincerely,

**Comment by lpasquali** [comment]
Hello, since yesterday I kept apache up, as it responds correctly when called directly.
In the mean while  I'm debugging a  bit and I found that error 400 for me is due to this error in an header which information is filled in by nginx address XXX.XXX.XXX.XXX is the ip address from which my desktop appears on the internet while YYY.YYY.YYY.YYY is the address of the cPanel vhost I'm contacting with curl:

2018/12/06 16:03:18 [info] 461557#461557: *4373 client sent too long header line: "X-Forwarded-For: XXX.XXX.XXX.XXX, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.YYY, YYY.YYY.YYY.Y.." while reading client request headers, client: YYY.YYY.YYY.YYY, server: localhost, request: "GET /monitor.php HTTP/1.1", host: "monitor-demo11.example.com"

I'm using this curl:

vanth:~ luca$  curl -I http://monitor-demo11.example.com/monitor.php
HTTP/1.1 400 Bad Request
Server: nginx
Date: Thu, 06 Dec 2018 14:06:53 GMT
Content-Type: text/html
Content-Length: 242
Connection: keep-alive

vanth:~ luca$

**Comment by lpasquali** [comment]
I was able to fix the problem working on Engintron configuration only:
/etc/nginx/conf.d/default.conf:

```
    # Set the port for HTTP proxying
    #set $PROXY_TO_PORT 80;
    set $PROXY_TO_PORT 8080;
```

this means the disable feature from WHM or console does not edit back correctly this value (should it maybe COMMENTED when Engintron is disabled, rather than reversing nginx on port 8080 to apache on port 80, which is of no use)?

**Comment by lpasquali** [comment]
I found the issue, it's a boung in engintron.sh, I thiunk you should not touch that value at all, but if you really want you should invert the two sed against PROXY_TO_PORT because:

- when you enable, there is port 80 not 8080 in your logic.
- when you disable, you like to put port 8080 to 80, it seems.

but in your code you do the opposite,
Cheers.

**Comment by omerekelly** [comment]
> # Set the port for HTTP proxying #set $PROXY_TO_PORT 80; set $PROXY_TO_PORT 8080;

This has fixed my issue #968

**Comment by mofolo** [comment]
Also having this issue where the PROYX_TO_PORT is not automatically changing when enabling engintron. Instead NGINX is just trying to listen to itself and failing to bind.

**Comment by fevangelou** [comment]
Fixed in v1.9.3.
